### PR TITLE
Add scalafix for contramap

### PR DESCRIPTION
**Pull request by LukaJCB**
This PR provides an automatic fix for #1850, which should provide a seamless upgrading experience :)

### CODE REVIEW

File: scalafix/rules/src/main/scala/fix/Cats_v1_0_0.scala [hunk 1]
```diff
@@ -73,6 +74,33 @@ case class RemoveCartesianBuilder(index: SemanticdbIndex)
   }
 }
 
+// ref: https://github.com/typelevel/cats/issues/1850
+case class ContraMapToLMap(index: SemanticdbIndex)
+  extends SemanticRule(index, "UseLMapInsteadOfContraMap") {
+
+  //val contraMap = Symbol("_root_.cats.functor.Contravariant.Ops.contramap")
```
Comment (kailuowang) [inline scalafix/rules/src/main/scala/fix/Cats_v1_0_0.scala]: dead code?

File: scalafix/rules/src/main/scala/fix/Cats_v1_0_0.scala [hunk 2]
```diff
@@ -73,6 +74,33 @@ case class RemoveCartesianBuilder(index: SemanticdbIndex)
...
+  override def fix(ctx: RuleCtx): Patch = {
+
+    val contraMatcher = SymbolMatcher.normalized(Symbol("_root_.cats.functor.Contravariant.Ops.`contramap`."))
+
+    val unApplyName = "catsUnapply2left"
+
+    ctx.tree.collect {
+      case Term.Apply(fun, _) =>
+        if (contraMatcher.matches(fun) &&
+          fun.children.headOption.flatMap(index.denotation).exists{ x => println(x.name == unApplyName); x.name == unApplyName }) {
```
Comment (kailuowang) [inline scalafix/rules/src/main/scala/fix/Cats_v1_0_0.scala]: Is this `println` intentionally left?

Comment (kailuowang) [reply scalafix/rules/src/main/scala/fix/Cats_v1_0_0.scala]: This denotation trick is nice

**Comment by kailuowang** [comment]
This is awesome! @LukaJCB . I would say this fixes #1850

**Comment by LukaJCB** [comment]
Btw, it might be good to test these with CI in the future :)

**Comment by kailuowang** [comment]
It's not? How did you test it locally? Is it hard to add it to travis?

**Comment by LukaJCB** [comment]
It's a different sbt project, let me see what I can do :)

File: .travis.yml [hunk 3]
```diff
@@ -30,6 +30,11 @@ matrix:
 before_install:
  - export PATH=${PATH}:./vendor/bundle
 
+before_script:
```
[the remark this answers is not in the document]
Comment (LukaJCB) [reply .travis.yml]: Agreed. Just want to test for now :)

**Comment by kailuowang** [comment]
Another PR added some rules that only applies to the yet to release RC1. So we need to let the output depends on a snapshot of head, need to publish the head snapshot locally first at CI.

**Comment by LukaJCB** [comment]
Seems like a good idea 👍  I'm not sure how to go forward with that though, you have any pointers?

**Comment by kailuowang** [comment]
@LukaJCB, change here https://github.com/typelevel/cats/blob/master/scalafix/build.sbt#L17
to point to `1.0.0-SNAPSHOT` instead of `1.0.0-MF`. 
Then in the travis build script add another step "sbt publishLocal" so it's 
```bash
sbt ;coreJVM/publishLocal;freeJVM/publishLocal
cd scalafix
sbt tests/test
cd ..
```

File: scalafix/rules/src/main/scala/fix/Cats_v1_0_0.scala [hunk 4]
```diff
@@ -73,6 +74,30 @@ case class RemoveCartesianBuilder(index: SemanticdbIndex)
...
+  extends SemanticRule(index, "UseLMapInsteadOfContraMap") {
+
+  override def fix(ctx: RuleCtx): Patch = {
+
+    val contraMatcher = SymbolMatcher.normalized(Symbol("_root_.cats.functor.Contravariant.Ops.`contramap`."))
+
+    val unApplyName = "catsUnapply2left"
+
+    ctx.tree.collect {
+      case Term.Apply(fun, _) =>
```
Comment (gabro) [inline scalafix/rules/src/main/scala/fix/Cats_v1_0_0.scala]: SymbolMatches have an `unapply` method, so you can use it like

```
case Term.Apply(contraMatcher(fun), _) =>
```

File: scalafix/rules/src/main/scala/fix/Cats_v1_0_0.scala [hunk 5]
```diff
@@ -73,6 +74,30 @@ case class RemoveCartesianBuilder(index: SemanticdbIndex)
...
+        if (contraMatcher.matches(fun) &&
+          fun.children.headOption.flatMap(index.denotation).exists(_.name == unApplyName )) {
+          fun.children.find(contraMatcher.matches).map(tree => ctx.replaceTree(tree, "lmap")).getOrElse(Patch.empty)
```
Comment (gabro) [inline scalafix/rules/src/main/scala/fix/Cats_v1_0_0.scala]: `.getOrElse(Patch.empty)` can become `.asPatch`

File: scalafix/rules/src/main/scala/fix/Cats_v1_0_0.scala [hunk 6]
```diff
@@ -73,6 +74,30 @@ case class RemoveCartesianBuilder(index: SemanticdbIndex)
...
+        } else {
+          Patch.empty
+        }
+      case _ => Patch.empty
```
Comment (gabro) [inline scalafix/rules/src/main/scala/fix/Cats_v1_0_0.scala]: no need for this case, since this is a `collect`

**Comment by LukaJCB** [comment]
I simplified a lot thanks to @gabro! 🎉

File: scripts/travis-publish.sh [hunk 7]
```diff
@@ -34,6 +35,11 @@ free_js="$sbt_cmd validateFreeJS"
 js="$core_js && $free_js && $kernel_js"
 jvm="$sbt_cmd coverage validateJVM coverageReport && codecov"
 
+sbt ;coreJVM/publishLocal;freeJVM/publishLocal
+cd scalafix
+sbt tests/test
```
Comment (edmundnoble) [inline scripts/travis-publish.sh]: I expected this to add a lot to build time, but apparently not!

**Comment by LukaJCB** [comment]
Merging with two sign offs :)

**Pull request merged**